packages/flutter_tools/test/widget_preview_scaffold.shard/widget_preview_scaffold/lib/src/widget_preview_scaffold_controller.dart

```diff
@@ -154,3 +161,77 @@ class WidgetPreviewScaffoldController {
     }
   }
 }
+
+/// A custom [WidgetInspectorService] responsible for routing navigation events
+/// to the IDE.
+class _WidgetPreviewScaffoldInspectorService with WidgetInspectorService {
+  _WidgetPreviewScaffoldInspectorService({required this.dtdServices}) {
+    WidgetInspectorService.instance = this;
+  }
+
+  final WidgetPreviewScaffoldDtdServices dtdServices;
+
+  // Keys used to specify the creation location of a widget when serializing a
+  // DiagnosticsNode to JSON. This location is used by the widget inspector
+  // to jump to the creation location of a selected widget.
+  static const kFile = 'fileUri';
+  static const kLine = 'line';
+  static const kColumn = 'column';
+
+  CodeLocation? _nextNavigationLocation;
+
+  @protected
+  @override
+  bool setSelection(Object? object, [String? groupName]) {
+    // The next navigation event sent to `postEvent` will be for this selection.
+    // Save the location of preview annotation applications so we can override
+    // the navigation target in `postEvent`.
+    if (object is PreviewWidgetElement) {
+      final previewData = (object.widget as PreviewWidget).preview;
+      _nextNavigationLocation = CodeLocation(
+        uri: previewData.scriptUri,
+        line: previewData.line,
+        column: previewData.column,
+      );
+    }
+    final result = super.setSelection(object, groupName);
```

Comment:
This seems a little awkward - would it be simpler to skip this call and just directly call `super.postEvent()` with this location instead? Or if we still need the other base code (setting the selection), what about making `setSelection` take an optional `Location`?

Reply:
It's definitely awkward, but `super.setSelection` needs to be called as there's a lot of other state that's updated as the result of a selection.

This code is called via a service extension, and DevTools doesn't know that there's anything special about the `PreviewWidget`, so we need to keep this logic contained within the previewer.